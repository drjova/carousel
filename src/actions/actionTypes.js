export const IMAGES_REQUEST = 'IMAGES_REQUEST';
export const IMAGES_SUCCESS = 'IMAGES_SUCCESS';
export const IMAGES_ERROR = 'IMAGES_ERROR';