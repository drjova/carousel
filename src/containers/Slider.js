import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import Loading from '../components/Loading';
import Slide from '../components/Slide';
import fetchImages from '../actions/images';

class Slider extends Component {
  componentDidMount() {
    this.props.dispatch(fetchImages());
  }

  renderImages() {
    const { images } = this.props;
    return images.map((image, index) => (
      <Slide image={image} width={300} height={300} key={index} />
    ));
  }

  renderNavigation() {}

  render() {
    const { loading } = this.props;
    if (loading) {
      return <div>Loading</div>;
    }
    return (
      <div>
        {this.renderNavigation()}
        {this.renderImages()}
      </div>
    );
  }
}

Slider.propTypes = {
  dispatch: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
  error: PropTypes.objectOf(PropTypes.any),
  images: PropTypes.arrayOf(PropTypes.string),
};

Slider.defaultProps = {
  images: [],
};

const mapStateToProps = state => ({
  loading: state.images.loading,
  images: state.images.data,
  error: state.images.error,
});
const dispatchToProps = dispatch => ({ dispatch });

export default connect(
  mapStateToProps,
  dispatchToProps
)(Slider);
