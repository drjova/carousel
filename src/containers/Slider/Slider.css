#slider {
  width: 100%;
}
#slider-wrapper {
  background-color: #fff;
  height: 350px;
  padding: 25px 0;
  overflow: hidden;
  display: flex;
  align-items: flex-start;
  justify-content: center;
}
#slider h1 {
  color: #4a739e;
  margin-bottom: 5px;
  font-size: 26px;
}
#slider-navigation {
  margin-top: 20px;
  display: flex;
  align-items: center;
  justify-content: center;
}
