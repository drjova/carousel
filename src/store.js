import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
/* eslint-disable import/no-extraneous-dependencies */
import { createLogger } from 'redux-logger';
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';
/* eslint-disable import/no-extraneous-dependencies */

import reducers from './reducers';
import http from './common/http';

export const thunkMiddleware = thunk.withExtraArgument(http);

const PROD_MIDDLEWARES = [
    thunkMiddleware,
];

const DEV_MIDDLEWARES = [...PROD_MIDDLEWARES, createLogger()];

const withMiddlewares = () => {
    if (process.env.NODE_ENV === 'production') {
      return applyMiddleware(...PROD_MIDDLEWARES);
    }
    return applyMiddleware(...DEV_MIDDLEWARES);
};

export const store = createStore(
    reducers,
    composeWithDevTools(withMiddlewares()),
);