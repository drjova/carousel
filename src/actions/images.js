import { IMAGES_ERROR, IMAGES_REQUEST, IMAGES_SUCCESS } from './actionTypes';
import API_URL from '../constants';

function requestImages() {
  return {
    type: IMAGES_REQUEST,
  };
}

function errorImages(payload) {
  return {
    type: IMAGES_ERROR,
    payload,
  };
}

function successImages(payload) {
  return {
    type: IMAGES_SUCCESS,
    payload,
  };
}

export default function fetchImages() {
  return async (dispatch, getState, http) => {
    dispatch(requestImages());
    try {
      const response = await http.get(API_URL);
      dispatch(successImages(response.data));
    } catch (error) {
      dispatch(errorImages(error.response && error.response.data));
    }
  };
}
