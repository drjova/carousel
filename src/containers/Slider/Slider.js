import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import Loading from '../../components/Loading';
import Slide from '../../components/Slide/Slide';
import Navigation from '../../components/Navigation/Navigation';
import fetchImages from '../../actions/images';
import './Slider.css';

class Slider extends Component {
  constructor(props) {
    super(props);
    this.state = {
      current: 0,
      total: 12,
    };
    this.clickNext = this.clickNext.bind(this);
    this.clickPrev = this.clickPrev.bind(this);
  }

  componentDidMount() {
    this.props.dispatch(fetchImages());
  }

  clickNext() {
    this.setState({
      current:
        this.state.current < this.state.total
          ? this.state.current + 1
          : this.state.total,
    });
  }

  clickPrev() {
    this.setState({
      current: this.state.current > 1 ? this.state.current - 1 : 0,
    });
  }

  renderImages() {
    const { images } = this.props;
    const displayImage = images[this.state.current];
    return <Slide image={displayImage} key={this.state.current} />;
  }

  renderTitle() {
    return <h1>Carousel Test</h1>;
  }

  render() {
    const { loading, images } = this.props;
    if (loading) {
      return <Loading loading />;
    }
    return (
      <div id="slider">
        {this.renderTitle()}
        <div id="slider-wrapper">{images && this.renderImages()}</div>
        <div id="slider-navigation">
          <Navigation
            onClickPrev={this.clickPrev}
            onClickNext={this.clickNext}
            total={this.state.total}
            current={this.state.current}
          />
        </div>
      </div>
    );
  }
}

Slider.propTypes = {
  dispatch: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
  error: PropTypes.objectOf(PropTypes.any),
  images: PropTypes.arrayOf(PropTypes.string),
};

Slider.defaultProps = {
  images: [],
};

const mapStateToProps = state => ({
  loading: state.images.loading,
  images: state.images.data,
  error: state.images.error,
});
const dispatchToProps = dispatch => ({ dispatch });

export default connect(
  mapStateToProps,
  dispatchToProps
)(Slider);
