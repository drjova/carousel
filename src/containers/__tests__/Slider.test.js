import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';

import { getStore, getStoreWithState } from '../../fixtures/store';
import { IMAGES_REQUEST } from '../../actions/actionTypes';
import Slider from '../Slider/Slider';

describe('Slider component', () => {
  it('dispatches fetch images record', () => {
    const store = getStoreWithState({
      images: {
        loading: false,
        data: ['url'],
      },
    });
    mount(
      <Provider store={store}>
        <Slider />
      </Provider>
    );
    const actions = store.getActions();
    const expectedAction = actions.find(
      action => action.type === IMAGES_REQUEST
    );
    expect(expectedAction).toBeDefined();
  });

  it('displays load when ``loading`` prop is true', () => {
    const store = getStoreWithState({
      images: {
        loading: true,
      },
    });
    const component = mount(
      <Provider store={store}>
        <Slider />
      </Provider>
    );
    expect(component).toMatchSnapshot();
  });
});
