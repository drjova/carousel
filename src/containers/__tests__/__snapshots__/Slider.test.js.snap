// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider component displays load when \`\`loading\`\` prop is true 1`] = `
<div>
  Loading
</div>
`;
